chainer_setup_build.py

```diff
@@ -268,3 +328,164 @@ def get_ext_modules():
 
     check_extensions(extensions)
     return extensions
+
+
+def _nvcc_gencode_options():
+    """Returns NVCC gencode options generated from NVCC command line help."""
+    help_string = subprocess.check_output(
+        ['nvcc', '--help']).decode('ascii').replace('\n', '')
+
+    arch_options = re.findall("'(compute_\d{2})'", help_string)
+    arch_options = sorted(list(set(arch_options)))
+    arch_options = list(filter(lambda x: x >= 'compute_30', arch_options))
+
+    code_options = re.findall("'(sm_\d{2})'", help_string)
+    code_options = sorted(list(set(code_options)))
+    code_options = list(filter(lambda x: x >= 'sm_30', code_options))
+
+    pairs = []
+    for code_option in code_options:
+        arch_option = code_option.replace('sm_', 'compute_')
+        if arch_option not in arch_options:
+            msg = "No virtual architecture corresponding to '{}'.".format(
+                code_option)
+            raise ValueError(msg)
+        pairs.append((arch_option, code_option))
+
+    gencode_options = []
+    for pair in pairs:
+        gencode_options.append('-gencode=arch={},code={}'.format(*pair))
+
+    return gencode_options
+
+
+class _UnixCCompiler(unixccompiler.UnixCCompiler):
+    src_extensions = list(unixccompiler.UnixCCompiler.src_extensions)
+    src_extensions.append('.cu')
+
+    def _compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts):
+        # For ordinal extension source files, just call the super class method.
+        if os.path.splitext(src)[1] != '.cu':
+            return unixccompiler.UnixCCompiler._compile(
+                self, obj, src, ext, cc_args, extra_postargs, pp_opts)
+
+        # For CUDA C source files, compile them with NVCC.
+        _compiler_so = self.compiler_so
+        try:
+            nvcc_path = build.get_nvcc_path()
+            self.set_executable('compiler_so', nvcc_path)
+
+            postargs = (_nvcc_gencode_options() +
+                        ['-O2', '--compiler-options="-fPIC"'])
+
+            cflags = ''
+            if 'CFLAGS' in os.environ:
```

Comment:
Do we need to make a new environment variable such as `NVFLAGS`? With this patch, users don't need to fix `CFLAGS` for cudnn.